probe-rs/src/probe/cmsisdap/mod.rs

```diff
@@ -1218,6 +1267,51 @@ impl SwoAccess for CmsisDap {
     }
 }
 
+impl RawJtagIo for CmsisDap {
+    fn state(&self) -> &JtagDriverState {
+        &self.jtag_driver_state
+    }
+    fn state_mut(&mut self) -> &mut JtagDriverState {
+        &mut self.jtag_driver_state
+    }
+    fn shift_bit(&mut self, tms: bool, tdi: bool, capture: bool) -> Result<(), DebugProbeError> {
+        self.jtag_driver_state.state.update(tms);
+        if let Some(sequence) = self.jtag_sequences.last_mut() {
+            if sequence.append(tms, tdi, capture).is_ok() {
+                return Ok(());
+            }
+        }
+        self.jtag_sequences.push(JtagSequence::new(
```

Comment:
Please extract this into a `JtagSequence::new_single_bit` constructor. That way, you should be able to initialize an empty sequence and append the first bit to it. This would mean that it is impossible for the first bit to go into the wrong position in the future by accident.